### PR TITLE
Changelog automation task

### PR DESCRIPTION
**Pull request by noud-github**
add Changelog automation task that can be used to update the changelog of a repo without user interaction using the yoast-cli as a tool do retreave the used data from github.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1889,6 +1890,184 @@ grunt.initConfig( {
...
+        taskName: {      // The name of your task.
+            options: {}, // Task-specific options.
+            src: {},     // Files to target.
+        }
+    }
+} )
+```
+
+#### Options
+##### useEditDistanceComapair
```
Comment (herregroen) [inline README.md]:
```suggestion
##### useEditDistanceCompare
```

File: README.md [hunk 2]
```diff
@@ -1889,6 +1890,184 @@ grunt.initConfig( {
...
+Type: `Boolean`  
+Default: `false`
+
+Setting this to `true` allows the deletion duplicate line items with a Distance Comapair valu higher than 90.
```
Comment (herregroen) [inline README.md]:
```suggestion
Setting this to `true` allows the deletion duplicate line items with a Distance Compare value higher than 90.
```

File: config/update-changelog-with-latest-pr-texts.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// Custom task
+module.exports = {
+	options: {
+		useEditDistanceComapair: true,
```
Comment (herregroen) [inline config/update-changelog-with-latest-pr-texts.js]:
```suggestion
		useEditDistanceCompare: true,
```

File: lib/get-changelog-using-wiki.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+};
+
+console.log (keyToSend);
+const { spawn } = require('child_process');
+const { exit } = require('process');
+const yoastcli = spawn('/tmp/Wiki/yoast-cli/yoast-cli.php', ['changelog:create'], {cwd:"/tmp/Wiki/yoast-cli/"});
+// set timeout to stop the script it input is not as expected for some reason.
+const timeout = setTimeout(function(){ yoastcli.kill()}, 30000);
+
+yoastcli.stdout.on('data', (data) => {
```
Comment (herregroen) [inline lib/get-changelog-using-wiki.js]: Code here currently doesn't match the Yoast code style, we should plan in a short meeting to setup VSCode to automatically apply that codestyle.

File: README.md [hunk 5]
```diff
@@ -1889,6 +1890,184 @@ grunt.initConfig( {
...
 
 
+
+
+
+### The `update-changelog-with-latest-pr-texts` task
+#### Using our configuration
+We implement the following configuration:
+- `options`
+    - The `useEditDistanceComapair` value is set to `true`.
```
Comment (herregroen) [inline README.md:1900]:
```suggestion
    - The `useEditDistanceCompare` value is set to `true`.
```

File: README.md [hunk 6]
```diff
@@ -1889,6 +1890,184 @@ grunt.initConfig( {
...
+            matchChangelogHeader:  /[=]= Changelog ==\n\n/ig,
+            newHeadertemplate: "== Changelog ==\n\n" +"= " + "VERSIONNUMBER" + " =\nRelease Date: " + "DATESTRING"  + "\n\n",
+            matchCorrectHeader: "= " + "VERSIONNUMBER" + "(.|\\n)*?\\n(?=(\\w\+?:\\n|= \\d+[\.\\d]+ =|= Earlier versions =))",
+            matchCorrectLines: "= " + "VERSIONNUMBER" + "(.|\\n)*?(?=(= \\d+[\.\\d]+ =|= Earlier versions =))",
+            matchCleanedChangelog: "= " + "VERSIONNUMBER" + "(.|\\n)*= Earlier versions =",
+            replaceCleanedChangelog: "= Earlier versions =",
+            pluginSlug: "wordpress-seo",
+            defaultChangelogEntrys: "",
+            useANewLineAfterHeader: true,
+            useEditDistanceComapair: true,
```
Comment (herregroen) [inline README.md:2027]:
```suggestion
            useEditDistanceCompare: true,
```

File: tasks/update-changelog-with-latest-pr-texts.js [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+ * 
+ * @get cleanChangelog
+ * @retruns {Object} multiline string
+ * 
+ */
+class ChangelogBuilder {
+	constructor(grunt , changelogIn, useEditDistanceComapair = false, useANewLineAfterHeader = true, pluginSlug) {
+		this.ChangelogMap = new Map();
+		this.grunt = grunt;
+		this.useEditDistanceComapair = useEditDistanceComapair;
```
Comment (herregroen) [inline tasks/update-changelog-with-latest-pr-texts.js]:
```suggestion
		this.useEditDistanceCompare = useEditDistanceCompare;
```

File: tasks/update-changelog-with-latest-pr-texts.js [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+		const key = `${value.match(new RegExp(  "[ a-zA-Z]+:" ))}`;
+		const lines = value.match(new RegExp( "(?<=\n)\\*([\n]|.)+?(?=\Z|\n\n|\n\\*|\n$)", "gm" ));
+		if (this.ChangelogMap.has(key)) {
+			this.ChangelogMap.get(key).append(lines);
+		} else {
+			const uniqueLines = new Unique(this.grunt);
+			uniqueLines.append(lines);
+			this.ChangelogMap.set(key, uniqueLines);
+		};
+		if (this.useEditDistanceComapair) {
```
Comment (herregroen) [inline tasks/update-changelog-with-latest-pr-texts.js]:
```suggestion
		if (this.useEditDistanceCompare) {
```

File: tasks/update-changelog-with-latest-pr-texts.js [hunk 9]
```diff
@@ -0,0 +1,371 @@
...
+ * @param {Object} grunt The grunt helper object.
+ * @returns {void}
+ */
+module.exports = function( grunt ) {
+	grunt.registerMultiTask(
+		"update-changelog-with-latest-pr-texts",
+		"Prompts the user for the changelog entries and updates the changelog entry in a file specified.",
+		function() {
+			let options = this.options( {
+				useEditDistanceComapair: false,
```
Comment (herregroen) [inline tasks/update-changelog-with-latest-pr-texts.js]:
```suggestion
				useEditDistanceCompare: false,
```

File: tasks/update-changelog-with-latest-pr-texts.js [hunk 10]
```diff
@@ -0,0 +1,371 @@
...
+
+					// Update the grunt reference to the changelog.
+					grunt.option( "changelog", changelog );
+
+					// Write changes to the file.
+					grunt.file.write( options.readmeFile, changelog );
+				}
+			}
+
+			const changelogBuilder = new ChangelogBuilder(grunt, null , options.useEditDistanceComapair,options.useANewLineAfterHeader , options.pluginSlug);
```
Comment (herregroen) [inline tasks/update-changelog-with-latest-pr-texts.js]:
```suggestion
			const changelogBuilder = new ChangelogBuilder(grunt, null , options.useEditDistanceCompare,options.useANewLineAfterHeader , options.pluginSlug);
```

File: test/update-changelog-1-test.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+					matchChangelogHeader:  /[=]= Changelog ==\n\n/ig,
+					newHeadertemplate: "== Changelog ==\n\n" +"= " + "VERSIONNUMBER" + " =\nRelease Date: " + "DATESTRING"  + "\n\n",
+					matchCorrectHeader: "= " + "VERSIONNUMBER" + "(.|\\n)*?\\n(?=(\\w\+?:\\n|= \\d+[\.\\d]+ =|= Earlier versions =))",
+					matchCorrectLines: "= " + "VERSIONNUMBER" + "(.|\\n)*?(?=(= \\d+[\.\\d]+ =|= Earlier versions =))",
+					matchCleanedChangelog: "= " + "VERSIONNUMBER" + "(.|\\n)*= Earlier versions =",
+					replaceCleanedChangelog: "= Earlier versions =",
+					pluginSlug: "wordpress-seo",
+					defaultChangelogEntrys: "",
+					useANewLineAfterHeader: true,
+					useEditDistanceComapair: true,
```
Comment (herregroen) [inline test/update-changelog-1-test.js]:
```suggestion
					useEditDistanceCompare: true,
```

File: test/update-changelog-2-test.js [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+					matchChangelogHeader:  /^/ig,
+					newHeadertemplate: "### " + "VERSIONNUMBER" + ": " + "DATESTRING"  + "\n",
+					matchCorrectLines: "### " + "VERSIONNUMBER" + "(.|\\n)*?(?=(### \\d+[\.\\d]+\: |$))",
+					matchCorrectHeader: "### " + "VERSIONNUMBER" + "(.|\\n)*?\\n(?=(\\w\+?:\\n|### \\d+[\.\\d]+\: |$))",
+					matchCleanedChangelog: "### " + "VERSIONNUMBER" + "(.|\\n)*$",
+					replaceCleanedChangelog: "",
+					pluginSlug: "wordpress-seo-premium",
+					defaultChangelogEntrys: "Other:\n* Includes every change in Yoast SEO core " + "VERSIONNUMBER" + ". See the [core changelog](https://wordpress.org/plugins/wordpress-seo/#developers).\n",
+					useANewLineAfterHeader: false,
+					useEditDistanceComapair: true,
```
Comment (herregroen) [inline test/update-changelog-2-test.js]:
```suggestion
					useEditDistanceCompare: true,
```

File: test/update-changelog-3-test.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+					matchChangelogHeader:  /[=]= Changelog ==\n\n/ig,
+					newHeadertemplate: "== Changelog ==\n\n" +"= " + "VERSIONNUMBER" + " =\nRelease Date: " + "DATESTRING"  + "\n\n",
+					matchCorrectHeader: "= " + "VERSIONNUMBER" + "(.|\\n)*?\\n(?=(\\w\+?:\\n|= \\d+[\.\\d]+ =|= Earlier versions =))",
+					matchCorrectLines: "= " + "VERSIONNUMBER" + "(.|\\n)*?(?=(= \\d+[\.\\d]+ =|= Earlier versions =))",
+					matchCleanedChangelog: "= " + "VERSIONNUMBER" + "(.|\\n)*= Earlier versions =",
+					replaceCleanedChangelog: "= Earlier versions =",
+					pluginSlug: "wordpress-seo",
+					defaultChangelogEntrys: "",
+					useANewLineAfterHeader: true,
+					useEditDistanceComapair: true,
```
Comment (herregroen) [inline test/update-changelog-3-test.js]:
```suggestion
					useEditDistanceCompare: true,
```

**Review by herregroen — CHANGES_REQUESTED** [review]

File: lib/get-changelog-using-wiki.js (hunk 4 above)
Comment (noud-github) [reply lib/get-changelog-using-wiki.js]: 👍

**Pull request merged**